Baseline Functionality Reset

janxious opened the pull request:
This does a lot. In no particular order:

* normalized metrics based on metric cheatsheet and other discussions
* configuration and on/off toggling added for all metrics
* tests for all pieces that are on by default
* nascent gem level differentiation for old v new
* travis setup changes to enable redis/memcached testing
* rack tests deserve a special mention

-----

Things it doesn't do:

* test configuration combinations
* fix all the issues with namespace weirdness in this gem
* rename to our decided on final name for this gem

-----

Additionally here is a list of things Jason and I talked about for a long while and their current status, and none of these may be handled by this PR:

* **DONE** need default logger
* **DONE** handle configuration missing
* *TODO* handle configuration bad
* **DONE** version/feature mismatch
* *TODO* handle errors in our coding
* *TODO* rename to Metrician
* **DONE** tests
* *TODO* installation tasks for rails/rake

janxious committed:
namespace the request timing middleware

janxious committed:
make the railtie initializer name tied to the tool instead of the framework

janxious committed:
fix metric listing syntax

janxious committed:
add configuration pieces

janxious committed:
add toggles from configuration to the request timing middleware and add 5xx status tracking as 'web.error'

janxious committed:
add toggles for the activerecord metric tracking

janxious committed:
add external service configuration toggles

janxious committed:
delayed job - use the correct module and fix it to work

janxious committed:
add configuration options to resque job reporting

janxious committed:
add configuration options to sidekiq job reporting

janxious committed:
add job reporting options to default configuration file

janxious committed:
unify memcached and redis cache tracking into the same metric namespace and add configuration options

janxious committed:
give logger fallback to the agent, which is stdout with WARN level by default

janxious committed:
add exception tracker configuration, and disable it

janxious committed:
change the specifics of command specific cache instrumentation's configuration

janxious committed:
fix the tests to reflect new world of simplified metrics

janxious committed:
add resque instrumentation test

janxious committed:
add tests for sidekiq

janxious committed:
setup travis to test older gems on 4.2 and newer on 5

janxious committed:
simplify agent setup in specs

janxious committed:
fix Gemfile(.lock) and add a check inside script/setup that the symlink isn't broken

janxious committed:
add redis cache spec

janxious committed:
add memcached cache tests for the dalli and memcached gems

janxious committed:
comment on why the memcached is as it is

janxious committed:
put active record tests in a describe block

janxious committed:
test external services

janxious committed:
add database and cache dependencies to travis setup

janxious committed:
memcached raises exceptions when a key has not been set

janxious committed:
use random value for get on redis cache test so that local dev having data isn't as much a problem

janxious committed:
add middleware tests

janxious committed:
colorize test run output

janxious committed:
add job queue error tests

janxious committed:
use tput instead of control characters for test output highlighting

jason-o-matic committed:
Take a stab at making config language friendlier

Our goal is no impact, so I want the config language to lean that way as well. I'm open to discussion!

jason-o-matic committed:
Clarify cache config comment

Before I read `command_specific` below I thought the previous comment indicated `command` would measure per-command.

jason-o-matic commented on lib/delayed_job/delayed_job_callbacks.rb:
```diff
@@ -8,20 +8,14 @@ class DelayedJobCallbacks < ::Delayed::Plugin
         ensure
           duration = Time.now - start
-          InstrumentalReporters.agent.gauge(job_metric_instrumentation_name(job), duration)
+          InstrumentalReporters.gauge("queue.process", duration) if InstrumentalReporters.configuration[:queue][:process][:enabled]
```
This isn't a big deal, but how would you feel about having a superclass or module on this class that had a gauge method and a config method, so this code looked something like this:
```
gauge("queue.process") if config[:process][:enabled]
```

janxious replied:
We might chain off something similarly to how `Instrumental::Agent` becomes `I`, but as we discussed, in the various pieces that are monkey-patching, defining helper methods on the object being patched is kind of bogus.

jason-o-matic commented on lib/instrumental/configuration.rb:
```diff
@@ -0,0 +1,17 @@
+
+    def self.location
+      ENV["METRICIAN_CONFIG"] ||
+        File.join(Dir.pwd, "config", "metrician.yaml")
```
I think we might want to expand the number of default locations to look for a config since it might be common to load the gem when you're not in the important directory.

Maybe the full list is:
1. ENV METRICIAN_CONFIG
2. ENV RAILS_ROOT
3. ENV RACK_ROOT
4. Home dir
5. `Dir.pwd`

jason-o-matic commented on lib/resque/resque_plugin.rb:
```diff
@@ -28,7 +29,7 @@ module ResqueHelper
       # remove all #, ?, !, etc. as well as runs of . and ending .'s
       name = job.to_s.gsub(/[^\w]+/, ".").gsub(/\.+$/, "")
-      "jobs.#{name}"
+      "queue.#{name}"
```
After some discussion we've decided to head towards this:
```
jobs.run
jobs.queue.<queue_name>
jobs.task.<job_name>
jobs.queue.<queue_name>.task.<job_class>
```

jason-o-matic commented on spec/instrumental_reporters_spec.rb:
```diff
@@ -5,29 +5,212 @@
+        # typically instrumental reporters would be loaded in an initalizer
+        # and this _extend_ could be done inside the job itself, but here
+        # we are in a weird situation.
+        TestResqueJob.send(:extend, Instrumental::ResquePlugin)
```
Is this needed because `InstrumentalReporters.activate` was called in a different test before `resque` was required?

I guess it seems like if the above line is necessary something with `activate` isn't working the way I expect.

EDIT:
I dug into this while making the lib loading changes and indeed, it does not work how I thought it did.

I thought we automatically hooked into all resque jobs, but that's not the case. Resque jobs are generally just bare classes that have a `perform` class method, so there isn't a super obvious place to hook in.

I think we should update our thing to do something like honeybadger to hook in automatically:
https://github.com/honeybadger-io/honeybadger-ruby/blob/master/lib/honeybadger/plugins/resque.rb#L38-L47

It works by hooking into [`Resque::Job#payload_class`](https://github.com/resque/resque/blob/master/lib/resque/job.rb#L208-L210).

janxious replied:
Agreed re: `payload_class`. Pretty neat.

jason-o-matic commented on spec/instrumental_reporters_spec.rb:
```diff
@@ -5,29 +5,212 @@
+      begin
+        require "memcached"
+        return Memcached.new("localhost:11211")
+      rescue LoadError
```
Gemika has a way to query your test env, which might make this prettier:
https://github.com/makandra/gemika#bridging-incompatibilities-between-dependency-sets

I think we can check to see if memcached (or dalli) is loaded, and test each one.

Now that I think about it, maybe it'd be better to loop over the list of loaded cache clients and test each one, in case you haven't multiple loaded for different purposes.

jason-o-matic replied:
I made this change while I was working on the loading thing, but I'll leave this comment for discussion purposes.

janxious replied:
There's a case that my code covered that your does not, which is that we are in a gemset where neither gem exists. In the original it would have been vociferous, and the new code it will silently just not run any tests, I think.

jason-o-matic replied:
Indeed. I'll push a thing in a sec that should cover both cases.

janxious committed:
make the configuration load the default defined in the included yaml file if none is found in the ENV or the project using this gem

jason-o-matic committed:
Simplify sidekiq testing slightly

Clarified the necessity of the middleware hook. Also setup requiring
more like a regular app, which simplifies the require/includes.

jason-o-matic committed:
Simplify library requirement

This makes it work more like a regular app with bundle requiring things.

jason-o-matic committed:
Simplified cache testing

This change tests whatever's available, which simplifies the loading
code, and will cover the case where both client may be included at the
same time.

jason-o-matic committed:
Clarify test message

jason-o-matic committed:
Merge branch 'toggle-bobble' of github.com:Instrumental/instrumental_reporters into toggle-bobble

jason-o-matic reviewed (approved):
I left a few comments mostly for discussion, but this is looking great and I'm cool with merging it.

jason-o-matic commented on lib/instrumental/configuration.rb:
```diff
@@ -0,0 +1,33 @@
+    end
+
+    def self.app_location
+      File.join(Dir.pwd, "config", "metrician.yaml")
```
GitHub ate I comment I had in the previous review about config loading, so I will try to reproduce!

I'm not sure how careful everyone is about where they run their app from. I think sometimes we start rails processes from the deploy home dir instead of the app dir. So, I think we should expand the search list a little to include RAILS_ENV, RACK_ENV, and the current user's HOME dir.

I'm thinking this order:
1. ENV METRICIAN_CONFIG
2. RAILS_ENV
3. RACK_ENV
4. HOME
5. `Dir.pwd`
6. `gem_location`

Thoughts?

janxious replied:
I would probably skip HOME, but otherwise something incorporating those other pieces would be reasonable.

jason-o-matic committed:
Raise if no memcached clients can be found

Just in case we remove the gems accidentally or something, we'll know
instead of non-existent tests not failing.

janxious commented:
@jason-o-matic you planning on pushing anything new here?

jason-o-matic commented:
> @jason-o-matic you planning on pushing anything new here?

I might work on the stuff we've discussed here as time allows, but I don't have anything outstanding. Feel free to tackle anything we've discussed as well. I'm cool with merging any time.
